bot/exts/fun/user_events.py

```diff
@@ -0,0 +1,583 @@
+from functools import wraps
+from typing import Callable, Optional, Tuple
+
+from dateutil.parser import isoparse, parse
+from dateutil.relativedelta import relativedelta
+from discord import Embed, Member, Message, Reaction
+from discord.ext.commands import Cog, CommandInvokeError, Context, group, has_role
+
+from bot.api import ResponseCodeError
+from bot.bot import Bot
+from bot.constants import Channels, Guild, Roles
+from bot.utils.scheduling import Scheduler
+from bot.utils.time import humanize_delta
+
+log = logging.getLogger(__name__)
+
+EMOJIS = {
+    "check": "✅",
+    "cross": "❌"
+}
+
+DATE_PREFIX = {
+    1: 'st', 21: 'st', 31: 'st',
+    2: 'nd', 22: 'nd',
+    3: 'rd', 23: 'rd'
+}
+
+NOT_SCHEDULED = "Not scheduled"
+LIVE = "Live"
+
+
+def is_event_organizer(func: Callable) -> Callable:
+    """Check if the user is event organizer."""
+    @wraps(func)
+    async def wrapper(self: Cog, ctx: Context, event_name: str, *args, **kwargs) -> None:
+        user_event = await self.bot.api_client.get(f"bot/user-events/{event_name}")
+
+        if user_event["organizer"] != ctx.author.id:
+            await ctx.send(f"You are not the organizer of the event **{event_name}**.")
+            return
+
+        await func(self, ctx, event_name, *args, **kwargs)
+
+    return wrapper
+
+
+class UserEvents(Cog):
+    """Manage user events with the provided commands."""
+
+    def __init__(self, bot: Bot) -> None:
+        self.bot = bot
+        self.scheduler = Scheduler(self.__class__.__name__)
+
+        self.guild = None
+
+        # Load required channels.
+        self.user_event_coord_channel = None
+        self.user_event_announcement_channel = None
+        self.user_events_list_channel = None
+        self.user_event_voice_channel = None
+
+        # Load required roles.
+        self.developers_role = None
+        self.user_event_ongoing_role = None
+
+        self.bot.loop.create_task(self.load_required_assets())
+        self.bot.loop.create_task(self.restart_event_reminders())
+
+    def cog_unload(self) -> None:
+        """Cancel scheduled tasks."""
+        self.scheduler.cancel_all()
+
+    async def load_required_assets(self) -> None:
+        """Load discord guild components required by this cog."""
+        await self.bot.wait_until_guild_available()
+
+        # Get guild
+        self.guild = self.bot.get_guild(Guild.id)
+
+        # Load required channels.
+        self.user_event_coord_channel = self.bot.get_channel(Channels.user_event_coordinators)
+        self.user_event_announcement_channel = self.bot.get_channel(Channels.user_event_announcements)
+        self.user_events_list_channel = self.bot.get_channel(Channels.user_event_list)
+        self.user_event_voice_channel = self.bot.get_channel(Channels.user_event_voice)
+
+        # Load required roles.
+        self.developers_role = self.guild.get_role(Roles.verified)
+        self.user_event_ongoing_role = self.guild.get_role(Roles.user_event_ongoing)
```

Comment:
Use here `get or fetch` pattern. When some channel/role is not found, this should unload cog and show warning about it, because everything is required there. Also, use asyncio.Event to wait for items loading.

Reply:
Discordpy does not provide the fetch method for channels.
I am going for the following pattern:
- using properties like before
- add a custom property decorator that caches the value and unloads the cog if the property value is None.

Reply:
discord.py have `fetch_channel` method.

Reply:
oh, right, how did I miss that.

Reply:
what if I check if `get_channel` or `get_role` returns None, then I unload the cog with appropriate logging instead of `fetch_channel` raising an error and unloading the cog?

Reply:
We must be ready that channels or role(s) is not in cache, so we need to fetch them.